server.js

```diff
@@ -785,41 +797,97 @@ cache(function (data, match, sendBadge, request) {
   });
 }));
 
+/**
+ * Registers a service in the shields.io system.
+ *
+ * `config` is an object containing these keys:
+ *  - name: Name of the service
+ *  - type: Type of the service (build, downloads, version, etc)
+ *  - uri: Regular expression to use for URIs for this service's badges
+ *  - examples: Example URIs for this service. These should use the format
+ *    specified in `uri`, and can be used to demonstrate how to use badges for
+ *    this service.
+ *  - handler: Asynchronous function to handle requests for this service. This
+ *    function takes two arguments:
+ *    1. Capture groups from the `uri` regular expression
+ *    2. A `request` method for sending cached HTTP requests to third-party
+ *       servers.
+ */
+function registerService(config) {
+  // Regular expressions treat "/" specially, so we need to escape them
+  const escapedPath = config.uri.replace(/\//g, '\\/');
+  const fullRegex = '^' + escapedPath + '.(svg|png|gif|jpg|json)$';
+
+  camp.route(new RegExp(fullRegex),
+  cache(function(data, match, sendBadge, request) {
+    // Assumes the final capture group is the extension
+    const format = match.pop();
+    // Discard the first capture group as it's the entire match
+    const params = match.slice(1);
+
+    const badgeData = getBadgeData(config.type, data);
+    config.handler(params, request.asPromise)
+      .then(serviceData => {
+        let text = badgeData.text;
+        if (serviceData.text) {
+          text[1] = serviceData.text;
+        }
+        Object.assign(badgeData, serviceData);
+        badgeData.text = text;
+        sendBadge(format, badgeData);
+      })
+      .catch(error => {
+        console.error(error);
+        const text = badgeData.text;
+        text[1] = 'error';
+        badgeData.text = text;
+        sendBadge(format, badgeData);
+      });
+  }));
+}
+
 // AppVeyor CI integration.
-camp.route(/^\/appveyor\/ci\/([^\/]+\/[^\/]+)(?:\/(.+))?\.(svg|png|gif|jpg|json)$/,
-cache(function(data, match, sendBadge, request) {
-  var repo = match[1];  // eg, `gruntjs/grunt`.
-  var branch = match[2];
-  var format = match[3];
-  var apiUrl = 'https://ci.appveyor.com/api/projects/' + repo;
-  if (branch != null) {
-    apiUrl += '/branch/' + branch;
-  }
-  var badgeData = getBadgeData('build', data);
-  request(apiUrl, { headers: { 'Accept': 'application/json' } }, function(err, res, buffer) {
-    if (err != null) {
-      badgeData.text[1] = 'inaccessible';
-      sendBadge(format, badgeData);
+registerService({
+  name: 'AppVeyor',
+  type: 'build',
+  uri: '/appveyor/ci/([^/]+/[^/]+)(?:/(.+))?',
+  examples: [
+    {uri: '/appveyor/ci/gruntjs/grunt'},
+    {
+      name: 'Branch',
+      params: '/appveyor/ci/gruntjs/grunt/master',
+    },
+  ],
+  handler: async ([repo, branch], request) => {
```

Comment:
>  To avoid mistakes when code is copied and pasted, and when capture groups are added and removed

Tests should catch breakages like that though, right? 😄 

I really really wish JavaScript had named capture groups, then the regex could look like:
```
/appveyor/ci/(?<repo>[^/]+/[^/]+)(?<branch>:/(.+))?
```
And matches would be an object with a `repo` and `branch` properties.

(I'm using .NET named capture group syntax for this example)

We could import XRegExp which would allow us to use named capture groups, if you like (fun side note - The developer of XRegExp is my manager at Facebook 😛 ). Alternatively, we could just wait until we refactor the URI parsing, which could also fix how this is handled.

I do like your idea though (specifying names of the capture groups as an array. I think I'll go with that for now, I think it's sufficiently clean.